<center> <h1>Assignment 1</h1> </center>

The following code loads the nba dataset (nba.csv) from an online repository into a dataframe named "nba".

The columns are organized as "Name",	"Team",	"Number",	"Position",	"Age",	"Height",	"Weight",	"College",	"Salary". Use this dataset to answer the questions below. You can create as many cells as you desire for all solutions.


In [1]:
import pandas as pd
nba = pd.read_csv("https://media.geeksforgeeks.org/wp-content/uploads/nba.csv") 
nba.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

### Question 1: 
Create a new dataframe named "nba_new" by adding a new column named "College Degree" right before the column "College". Fill the rows of this column with "True" if there is a college name entry under the column "College" for the associated row. Make sure the new column named "College Degree" is located right before the column "College".<br/>

### Solution 1:

In [2]:
nba.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


It looks like there are 458 rows in the dataframe but each column only has 457 non-null values.   
Let's see whats going on...

In [3]:
nba[nba["Name"].isna()]

Name Team  Number Position  Age Height  Weight College  Salary
457  NaN  NaN     NaN      NaN  NaN    NaN     NaN     NaN     NaN

The row 457 is all `NaN` values. I'll remove it when creating `nba_new` dataframe.

In [4]:
nba_new = nba.drop(457)

In [5]:
nba_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 457 entries, 0 to 456
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 35.7+ KB


In [6]:
nba_new[nba_new["Name"].isna()] # make sure it is gone

Empty DataFrame
Columns: [Name, Team, Number, Position, Age, Height, Weight, College, Salary]
Index: []

In [7]:
nba_new.tail()

Name       Team  Number Position   Age Height  Weight   College  \
452    Trey Lyles  Utah Jazz    41.0       PF  20.0   6-10   234.0  Kentucky   
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0    Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0       NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0       NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0    Kansas   

        Salary  
452  2239800.0  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0

The row 457 has been removed. I'll make sure `nba` is unchanged.

In [8]:
nba.info()  # make sure nba is unchanged

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 458 entries, 0 to 457
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      457 non-null    object 
 1   Team      457 non-null    object 
 2   Number    457 non-null    float64
 3   Position  457 non-null    object 
 4   Age       457 non-null    float64
 5   Height    457 non-null    object 
 6   Weight    457 non-null    float64
 7   College   373 non-null    object 
 8   Salary    446 non-null    float64
dtypes: float64(4), object(5)
memory usage: 32.3+ KB


In [9]:
nba.tail()  # make sure nba is unchanged

Name       Team  Number Position   Age Height  Weight College  \
453  Shelvin Mack  Utah Jazz     8.0       PG  26.0    6-3   203.0  Butler   
454     Raul Neto  Utah Jazz    25.0       PG  24.0    6-1   179.0     NaN   
455  Tibor Pleiss  Utah Jazz    21.0        C  26.0    7-3   256.0     NaN   
456   Jeff Withey  Utah Jazz    24.0        C  26.0    7-0   231.0  Kansas   
457           NaN        NaN     NaN      NaN   NaN    NaN     NaN     NaN   

        Salary  
453  2433333.0  
454   900000.0  
455  2900000.0  
456   947276.0  
457        NaN

 Now, to continue with the analysis.

In [10]:
col_deg = ~(nba_new['College'].isna())
col_deg

0       True
1       True
2       True
3       True
4      False
       ...  
452     True
453     True
454    False
455    False
456     True
Name: College, Length: 457, dtype: bool

In [11]:
nba_new['College_degree'] = col_deg  # create the new col

In [12]:
nba_new.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  College_degree  
0              Texas  7730337.0            True  
1          Marquette  6796117.0            True  
2  Boston University        NaN            True  
3      Georgia State  1148640.0            True  
4                NaN  5000000.0           False

In [13]:
mycol = list(range(9)) # create a list of the desired col order
mycol.insert(7,9)      # insert col 9 where I want it before 'College' at index 7
mycol

[0, 1, 2, 3, 4, 5, 6, 9, 7, 8]

In [14]:
nba_new = nba_new.iloc[:,mycol]  # rearrange col with slice notation using the list I created above
nba_new.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

   College_degree            College     Salary  
0            True              Texas  7730337.0  
1            True          Marquette  6796117.0  
2            True  Boston University        NaN  
3            True      Georgia State  1148640.0  
4           False                NaN  5000000.0

### Question 2:
Report the first 5 rows of nba_new dataframe.<br/>

### Solution 2:

In [15]:
nba_new.head(5)

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

   College_degree            College     Salary  
0            True              Texas  7730337.0  
1            True          Marquette  6796117.0  
2            True  Boston University        NaN  
3            True      Georgia State  1148640.0  
4           False                NaN  5000000.0

### Question 3:
Make sure the new column is of data type boolean and report it is actually boolean.<br/>


### Solution 3:

In [16]:
nba_new['College_degree'].dtype

dtype('bool')

#### *Ans: data in col "College_degree" is indeed of type bool.*

### Question 4:
Create a list that holds the names of the players who do not have any college degree (or no university name entry given under "College" column in the data set) through using a for loop and if statement. Report only the number of those players.<br/>

### Solution 4:

In [17]:
type(nba_new['Name'])

pandas.core.series.Series

In [18]:
nba_new['Name'][0]

'Avery Bradley'

In [19]:
deg = []
nodeg = []
for i in range(len(nba_new['Name'])):
    if nba_new['College_degree'][i]:
        deg.append(nba_new['Name'][i])
    else:
        nodeg.append(nba_new['Name'][i])

In [20]:
len(nodeg)

84

In [21]:
len(nba_new['Name']) - sum(nba_new['College_degree']) # double check previous answer

84

In [23]:
nba_new[nba_new['College'].isna()]['Name'].count()  # answer with one line in pandas

84

#### *Ans: There are 84 players without a college degree*

### Question 5:
Create two dataframes named "no_college_degree" and "with_college_degree". no_college_degree dataframe should include two columns one for names and one for salaries of those players who do not hold a college degree. Likewise, with_college_degree dataframe should include two columns one for names and one for salaries of those players who do not hold a college degree. Report the first 5 rows of each dataframe. Use for loop and if statement in your solution.<br/>

### Solution 5:

In [24]:
no_cd = []
with_cd = []
index_no = []
index_w = []

for i in range(len(nba_new['Name'])):
    if nba_new['College_degree'][i]:
        with_cd.append((nba_new['Name'][i], nba_new['Salary'][i]))
        index_w.append(i)
    else:
        no_cd.append((nba_new['Name'][i], nba_new['Salary'][i]))
        index_no.append(i)

In [25]:
print("no_cd:", len(no_cd), " index_no:", len(index_no))
print("with_cd:", len(with_cd), " index_w:", len(index_w))

no_cd: 84  index_no: 84
with_cd: 373  index_w: 373


In [26]:
no_college_degree = pd.DataFrame(no_cd, index = index_no, columns = ["Name","Salary"])  # new df for players w/o college degree
with_college_degree = pd.DataFrame(with_cd, index = index_w, columns = ["Name","Salary"]) # new df for palyers with college degree

In [27]:
no_college_degree

Name      Salary
4             Jonas Jerebko   5000000.0
5              Amir Johnson  12000000.0
15         Bojan Bogdanovic   3425510.0
20           Sergey Karasev   1599840.0
32   Thanasis Antetokounmpo     30888.0
..                      ...         ...
445              Dante Exum   3777720.0
447             Rudy Gobert   1175880.0
450              Joe Ingles   2050000.0
454               Raul Neto    900000.0
455            Tibor Pleiss   2900000.0

[84 rows x 2 columns]

In [28]:
with_college_degree

Name     Salary
0    Avery Bradley  7730337.0
1      Jae Crowder  6796117.0
2     John Holland        NaN
3      R.J. Hunter  1148640.0
6    Jordan Mickey  1170960.0
..             ...        ...
449    Rodney Hood  1348440.0
451  Chris Johnson   981348.0
452     Trey Lyles  2239800.0
453   Shelvin Mack  2433333.0
456    Jeff Withey   947276.0

[373 rows x 2 columns]

Below is the solution using Pandas

In [29]:
no_degree = nba_new[~nba_new['College_degree']].iloc[:,[0,-1]]   # new df for players w/o college degree
with_degree = nba_new[nba_new['College_degree']].iloc[:,[0,-1]]  # new df for palyers with college degree

### Question 6:
Report the average salary of players with college and no college degrees.<br/>


### Solution 6:

In [30]:
no_degree['Salary'].mean()

5829803.426829268

In [31]:
with_degree['Salary'].mean()

4620311.071428572

It looks like the average salary for players *without* a degree is greater (\$5,829,803.43) the average salary of players *with* a college degree (\$4,620,311.07)!

Below I try doing the same analysis using `.groupby` on original dataframe:

In [32]:
nba_new.groupby('College_degree').mean()['Salary']

College_degree
False    5.829803e+06
True     4.620311e+06
Name: Salary, dtype: float64

### Question 7:
Rename column named "Height" to "Height (feet-inches)". Add a new column named "Height (inches)" to the existing dataframe. The column should include values in total inches. Say, for example, if a player's height is 6-0, the new column should show it as 72. Add another column named "Tall" to the existing dataframe. This column should indicate a bool type value of True if the associated player height is greater than or equal to 80 inches. Otherwise, its value should be False. Use for loop and if statement in your solution.<br/>

### Solution 7:

In [33]:
inches = []

for h in nba_new['Height']:
    ht_str_list = h.split(sep='-')   #split into a list based on the "-" character
    tot = int(ht_str_list[0])*12 + int(ht_str_list[1])  # remember to convert to type int 
    inches.append(tot)

In [34]:
len(inches)  # make sure it is correct length

457

In [35]:
nba_new.loc[:,'Height_inches'] = inches

In [12]:
# nba_new["Height_inches"] = inches  # throws a `SettingWithCopyWarning` a better way to do it is above

# I'm not sure why I'm getting this message but the code still seems to be working...
# SettingWithCopyWarning: 
# A value is trying to be set on a copy of a slice from a DataFrame.
# Try using .loc[row_indexer,col_indexer] = value instead

# Try converting inches to a dataframe:
# nba_new["Height_inches"] = pd.DataFrame(inches)

In [36]:
nba_new.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

   College_degree            College     Salary  Height_inches  
0            True              Texas  7730337.0             74  
1            True          Marquette  6796117.0             78  
2            True  Boston University        NaN             77  
3            True      Georgia State  1148640.0             77  
4           False                NaN  5000000.0             82

In [38]:
tall = []

for i in nba_new['Height_inches']:
    if i >= 80:
        tall.append(True)
    else:
        tall.append(False)

In [39]:
nba_new.insert(10, "Tall", tall)

In [40]:
nba_new.insert(10, "Tall2", nba_new['Height_inches'] >= 80)  # practice using Pandas not for and if loops

In [41]:
nba_new

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

     College_degree            College     Salary  Tall2   Tall  Height_inches  
0              True              Texas  7730337.0  False  False             74  
1              True          Marquette  6796117.0  False  False             78  
2              True  Boston University        NaN  False  False             77  
3              True      Georgia State  1148640.0  False  False             77  
4             False                NaN  5000000.0   True   True             82  
..              ...                ...        ...    ...    ...            ...  
452            True           Kentucky  2239800.0   True   True             82  
453            True             Butler  2433333.0  False  False             75  
454           False                NaN   900000.0  False  False             73  
455           False                NaN  2900000.0   True   True             87  
456            True             Kansas   947276.0   True   True             84  

[457 rows x 13 columns]

In [42]:
del nba_new['Tall2']  # delete the practice col again
nba_new

Name            Team  Number Position   Age Height  Weight  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2     John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   
..             ...             ...     ...      ...   ...    ...     ...   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   6-10   234.0   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0    6-3   203.0   
454      Raul Neto       Utah Jazz    25.0       PG  24.0    6-1   179.0   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0    7-3   256.0   
456    Jeff Withey       Utah Jazz    24.0        C  26.0    7-0   231.0   

     College_degree            College     Salary   Tall  Height_inches  
0              True              Texas  7730337.0  False             74  
1              True          Marquette  6796117.0  False             78  
2              True  Boston University        NaN  False             77  
3              True      Georgia State  1148640.0  False             77  
4             False                NaN  5000000.0   True             82  
..              ...                ...        ...    ...            ...  
452            True           Kentucky  2239800.0   True             82  
453            True             Butler  2433333.0  False             75  
454           False                NaN   900000.0  False             73  
455           False                NaN  2900000.0   True             87  
456            True             Kansas   947276.0   True             84  

[457 rows x 12 columns]

### Question 8:
By using the dataframe with the new column "Tall", report the average salaries of players who are labeled as 'True' under Tall column. Also, report the average salaries of players who are labeled as 'False' under Tall column.<br/>

### Solution 8:

In [43]:
nba_new.groupby('Tall').mean()['Salary']

Tall
False    4.310129e+06
True     5.361100e+06
Name: Salary, dtype: float64

On average tall players (greater than 80 inches in height) get paid about \\$1 million more than shorter players.  
* Short average: \\$4,310,129  
* Tall average: \$5,361,100

### Question 9:
Add a new column named "Young" to the existing dataframe. This column should indicate a bool type value of True if the associated player age is less than 30. Otherwise, its value should be False. Use for loop and if statement in your solution.<br/>

### Solution 9:

In [44]:
young = []

for age in nba_new['Age']:
    if age < 30:
        young.append(True)
    else:
        young.append(False)

In [45]:
len(young)

457

In [46]:
nba_new.insert(5, "Young", young)

In [47]:
nba_new  # no old players in preview

Name            Team  Number Position   Age  Young Height  \
0    Avery Bradley  Boston Celtics     0.0       PG  25.0   True    6-2   
1      Jae Crowder  Boston Celtics    99.0       SF  25.0   True    6-6   
2     John Holland  Boston Celtics    30.0       SG  27.0   True    6-5   
3      R.J. Hunter  Boston Celtics    28.0       SG  22.0   True    6-5   
4    Jonas Jerebko  Boston Celtics     8.0       PF  29.0   True   6-10   
..             ...             ...     ...      ...   ...    ...    ...   
452     Trey Lyles       Utah Jazz    41.0       PF  20.0   True   6-10   
453   Shelvin Mack       Utah Jazz     8.0       PG  26.0   True    6-3   
454      Raul Neto       Utah Jazz    25.0       PG  24.0   True    6-1   
455   Tibor Pleiss       Utah Jazz    21.0        C  26.0   True    7-3   
456    Jeff Withey       Utah Jazz    24.0        C  26.0   True    7-0   

     Weight  College_degree            College     Salary   Tall  \
0     180.0            True              Texas  7730337.0  False   
1     235.0            True          Marquette  6796117.0  False   
2     205.0            True  Boston University        NaN  False   
3     185.0            True      Georgia State  1148640.0  False   
4     231.0           False                NaN  5000000.0   True   
..      ...             ...                ...        ...    ...   
452   234.0            True           Kentucky  2239800.0   True   
453   203.0            True             Butler  2433333.0  False   
454   179.0           False                NaN   900000.0  False   
455   256.0           False                NaN  2900000.0   True   
456   231.0            True             Kansas   947276.0   True   

     Height_inches  
0               74  
1               78  
2               77  
3               77  
4               82  
..             ...  
452             82  
453             75  
454             73  
455             87  
456             84  

[457 rows x 13 columns]

In [48]:
nba_new[~nba_new['Young']]  # trying to make sure old players are False

Name                    Team  Number Position   Age  Young  \
19      Jarrett Jack           Brooklyn Nets     2.0       PG  32.0  False   
30     Arron Afflalo         New York Knicks     4.0       SG  30.0  False   
31      Lou Amundson         New York Knicks    17.0       PF  33.0  False   
33   Carmelo Anthony         New York Knicks     7.0       SF  32.0  False   
34     Jose Calderon         New York Knicks     3.0       PG  34.0  False   
..               ...                     ...     ...      ...   ...    ...   
415       Randy Foye   Oklahoma City Thunder     6.0       SG  32.0  False   
420    Nazr Mohammed   Oklahoma City Thunder    13.0        C  38.0  False   
421   Anthony Morrow   Oklahoma City Thunder     2.0       SG  30.0  False   
434      Chris Kaman  Portland Trail Blazers    35.0        C  34.0  False   
440    Brian Roberts  Portland Trail Blazers     2.0       PG  30.0  False   

    Height  Weight  College_degree           College      Salary   Tall  \
19     6-3   200.0            True      Georgia Tech   6300000.0  False   
30     6-5   210.0            True              UCLA   8000000.0  False   
31     6-9   220.0            True              UNLV   1635476.0   True   
33     6-8   240.0            True          Syracuse  22875000.0   True   
34     6-3   200.0           False               NaN   7402812.0  False   
..     ...     ...             ...               ...         ...    ...   
415    6-4   213.0            True         Villanova   3135000.0  False   
420   6-10   250.0            True          Kentucky    222888.0   True   
421    6-5   210.0            True      Georgia Tech   3344000.0  False   
434    7-0   265.0            True  Central Michigan   5016000.0   True   
440    6-1   173.0            True            Dayton   2854940.0  False   

     Height_inches  
19              75  
30              77  
31              81  
33              80  
34              75  
..             ...  
415             76  
420             82  
421             77  
434             84  
440             73  

[122 rows x 13 columns]

### Question 10:
By using the dataframe with the new column "Young", report the average salaries of players who are labeled as 'True' under Young column. Also, report the average salaries of players who are labeled as 'False' under Young column.<br/>

### Solution 10:

In [49]:
nba_new.groupby('Young').mean()['Salary']

Young
False    6.263216e+06
True     4.325732e+06
Name: Salary, dtype: float64

Young players get paid about \$2 million less than older players.
* Young average: \$4,325,732
* Old average: \$6,263,216